Remove timeouts from discover flows, fix desktop discovery

ryanclark opened the pull request:
## Fix desktop discovery

This adds a cache to the `useJoinToken` hook that caches the result of the join token against the resource kind. This is cleared only once a user leaves the entire flow, instead of the previous behaviour of clearing when the hook was unmounted.

This fixes the desktop discovery stage, as the cached result persists even though the hook has been unmounted between steps.

## Timer removal

This removes the concept of a script timing out in the discover workflows.

Instead, after 5 minutes of not being able to detect the relevant resource, we show a hint box that explains why we might not be seeing the resource, but we still let the user know we're still looking.

### Servers

<img width="1795" alt="node" src="https://user-images.githubusercontent.com/7922109/215143070-f90e0625-4380-4779-9158-b27e189fcf2b.png">

### Databases

<img width="1795" alt="database" src="https://user-images.githubusercontent.com/7922109/215143166-486ab84f-e81e-44b4-bdd8-acd3d33fd971.png">

### Kubernetes

For kubernetes, the flow has changed a little as we allowed the user to regenerate the command after editing the labels. As we don't actually need to regenerate the command, when editing the text fields the script value would change. To avoid this weirdness, I've made it so once you click on next, it disables the text fields.

Clicking on edit to change them will then hide the command, and once they click next again, they'll see the command again.

<img width="1795" alt="kubernetes" src="https://user-images.githubusercontent.com/7922109/215143194-ba29c5da-4777-4063-a8b5-75911161c317.png">

### Desktops

<img width="1795" alt="desktops" src="https://user-images.githubusercontent.com/7922109/215143222-b52c8894-b1f1-4483-a7ed-9c8b863da1b1.png">

<img width="1795" alt="discover-desktops" src="https://user-images.githubusercontent.com/7922109/215143262-da5d8ff8-2166-4f69-8598-56a421082279.png">

This fixes https://github.com/gravitational/teleport/issues/20149, https://github.com/gravitational/teleport/issues/20148 and https://github.com/gravitational/teleport/issues/19492.

ryanclark committed:
Allow for inputs to be disabled

ryanclark committed:
Remove timeouts from the discover workflows

ryanclark committed:
Correct the resource type to be cleared on unmount

ryanclark committed:
Update test snapshots

ibeckermayer commented on web/packages/teleport/src/Discover/Shared/useJoinToken.ts:
```diff
@@ -0,0 +1,101 @@
+      return {
+        joinToken: existing.response,
+        reloadJoinToken() {
+          joinTokenCache.delete(resourceKind);
+
+          rerender(c => c + 1);
```
```suggestion
          // Delete the cached token and force a rerender
          // so that this hook runs again and creates a new one.
          joinTokenCache.delete(resourceKind);
          rerender(c => c + 1);
```

ibeckermayer commented on web/packages/teleport/src/Discover/Shared/useJoinToken.ts:
```diff
@@ -0,0 +1,101 @@
+    throw existing.promise;
+  }
+
+  throw run().promise;
```
What happens with this on a re-render? Is `<Suspense>` smart enough to know to go back into its fallback state?

ryanclark replied:
On a re-render it'll go down the `if (existing)` route and return the existing promise

ibeckermayer commented on web/packages/teleport/src/Discover/Shared/useJoinToken.ts:
```diff
@@ -0,0 +1,101 @@
+  joinTokenCache.delete(resourceKind);
+}
+
+export function useJoinToken(
```
Thinking that we should create (and document in a developer section of the README) a naming convention that denotes this as a suspense hook (perhaps something like `useJoinTokenSuspender`). 

My reasoning is that `Suspense` is an experimental feature, and it's API is still unconfirmed and I wasn't able to find reliable documentation for it when initially googling. If I was new to the project, I would be totally baffled by this component.

That [example you linked me](https://codesandbox.io/s/frosty-hermann-bztrp?file=/src/fakeApi.js) is pretty good, although I wonder if there's also an official or unofficial documentation page that I missed.

ryanclark replied:
Sure, we can rename this

ibeckermayer commented on web/packages/teleport/src/Discover/Shared/usePoll.ts:
```diff
@@ -2,20 +2,17 @@ import { useEffect, useRef, useState } from 'react';
   enabled: boolean,
   interval = 1000
-) {
+): T | null {
```
Are you noticing a bug here wherein Typescript doesn't seem to pick up that this can return `null`? I notice that my editor (vscode) doesn't seem to pick it up, and if I take out `&& result` from this line

https://github.com/gravitational/teleport/blob/b20144dd090fd4128502c1cf2ac978f7de87d25f/web/packages/teleport/src/Discover/Desktop/DiscoverDesktops/DiscoverDesktops.tsx#L109

a run of `yarn type-check` doesn't complain at me

```
$ yarn type-check
yarn run v1.22.19
$ tsc --noEmit
✨  Done in 43.69s.
```

ryanclark replied:
Yeah I noticed this as well, it just returns `T`. I was just trying to be strict with typings by adding this 🤷

ibeckermayer replied:
Huh, seems like a ts bug. It might be worth adding explicit typings to places where the automatic typing is failing us and see if that helps.

ibeckermayer commented on web/packages/teleport/src/Discover/Shared/usePoll.ts:
```diff
@@ -24,16 +21,6 @@ export function usePoll<T>(
     }
   }, [callback, enabled, running]);
```
Why is `callback` needed in this sensitivity list?

ibeckermayer replied:
In fact, can we not simplify this whole hook to be

```typescript
export function usePoll<T>(
  callback: (signal: AbortSignal) => Promise<T | null>,
  enabled: boolean,
  interval = 1000
): T | null {
  const abortController = useRef(new AbortController());
  const [result, setResult] = useState<T | null>(null);

  useEffect(() => {
    if (enabled) {
      abortController.current = new AbortController();

      const id = window.setInterval(async () => {
        try {
          const result = await callback(abortController.current.signal);

          if (result) {
            clearInterval(id);
            setResult(result);
          }
          // eslint-disable-next-line no-empty
        } catch {}
      }, interval);

      return () => {
        clearInterval(id);
        abortController.current.abort();
      };
    } else {
      setResult(null);
    }
  }, [enabled, interval, callback]);

  return result;
}
```

kimlisa replied:
actually nvm, i think i can make use of the `enabled` flag

ryanclark replied:
Yeh I think so, I've updated it to use yours

kimlisa commented:
> Databases
> I've removed the generate command button from this flow. The script isn't dependent on the labels, so it doesn't make sense to hide it behind a button.

the script isn't dependent on labels, but the join token that we fetch right before generating script, is (it has a field for `suggested matcher labels`) use case is something like this:

1) create db step: create a database with labels `env: dev`
2) deploy service step: user wants to have this db service also pick up `env: prod` or `env: staging` for next time

i think the only case where we should not allow user to add more labels at deploy step, is when we default to asteriks.

Also, i don't think we handle the case where the token does expire (user takes a while trouble shooting or steps out for a while). Granted might not likely happen but if it does, the user will be stuck and have to start over?

ibeckermayer commented on web/packages/teleport/src/Discover/Shared/PingTeleportContext.tsx:
```diff
@@ -3,41 +3,37 @@ import React, { useCallback, useContext, useEffect, useState } from 'react';
 import { useTeleport } from 'teleport';
```
Looks like we can delete the `TODO` comment in this file (line 72-74)

ibeckermayer commented on web/packages/teleport/src/Discover/Server/DownloadScript/DownloadScript.tsx:
```diff
@@ -137,6 +107,60 @@ export function DownloadScript(props: AgentStepProps) {
+            </TextIcon>
+            We're still looking for your server
+          </Flex>
+        </Text>
+
+        <Text mb={3}>
+          There are a couple of possible reasons for why we haven't been able to
+          detect your server.
+        </Text>
+
+        <Text mb={1}>
+          - The command was not run on the server you were trying to add.
+        </Text>
+
+        <Text mb={3}>
+          - The Teleport SSH Service could not join this Teleport cluster. Check
+          the logs for errors by running <Mark>journalctl status teleport</Mark>
+          .
+        </Text>
+
+        <Text>
+          We'll continue to look for the server whilst you diagnose the issue.
+        </Text>
+      </HintBox>
+    );
+  } else {
+    hint = (
+      <WaitingInfo>
+        <TextIcon
+          css={`
+            white-space: pre;
+          `}
+        >
+          <Icons.Restore fontSize={4} />
+        </TextIcon>
+        After running the command above, we'll automatically detect your new
+        Teleport instance.
+      </WaitingInfo>
+    );
+  }
```
Something similar pops up multiple times, it looks like something we can parameterize with props (including the timeout) and hold in `web/packages/teleport/src/Discover/Shared`.

ryanclark replied:
I've simplified the usage of `HintBox` and moved the common hook behaviour into a new `useShowHint()`

ibeckermayer reviewed (changes requested):
Nice work getting rid of the context and keeping join token logic in a single hook

kimlisa reviewed (commented):
looks like stories broke from the story changes, and some story states will not be needed (if we ultimately not do certain things like, handling 4 hour timeout)

kimlisa commented on web/packages/teleport/src/Discover/Shared/usePoll.ts:
```diff
@@ -2,20 +2,17 @@ import { useEffect, useRef, useState } from 'react';
   const abortController = useRef(new AbortController());
 
   const [running, setRunning] = useState(false);
-  const [timedOut, setTimedOut] = useState(false);
```
actually nvm, i think i can make use of the `enabled` flag

ryanclark committed:
Rename useJoinToken to useJoinTokenSuspender

ryanclark committed:
Remove TODO

ryanclark committed:
Simplify usePoll

ryanclark commented on web/packages/teleport/src/Discover/Kubernetes/HelmChart/HelmChart.tsx:
```diff
@@ -45,27 +48,26 @@ import type { AgentStepProps } from '../../types';
-export default function Container(
-  props: AgentStepProps & { runJoinTokenPromise?: boolean }
-) {
+const SHOW_HINT_TIMEOUT = 1000 * 60 * 5; // 5 minutes
```
Done

ryanclark committed:
Simplify the hint box code, add a shared hook

ryanclark committed:
Add the generte command button back to db flow

ryanclark committed:
Store the joinToken in PingTeleportContext state instead of using the hook

ibeckermayer reviewed (approved):
LGTM once we've manually checked that everything is running as expected.

kimlisa committed:
Fix rebase diff mistake

kimlisa reviewed (approved):
i tested on my staging cloud cluster and everything works (i did not thoroughly test the desktop flow though, i just clicked through it but did not set anything up)

the only thing we need is to have a success state after discovering resources so it's more obvious something happened

kimlisa committed:
Fix bug with polling for result

- once we get the result, we set the result
  back to null, so we can't proceed to next step

kimlisa commented on web/packages/teleport/src/Discover/Shared/usePoll.ts:
```diff
@@ -7,9 +7,21 @@ export function usePoll<T>(
+
+    if (!enabled && running) {
+      setRunning(false);
+    }
```
i copy-pasta from your old code, but looking at this block again, it doesn't seem like i need it 🤔

ravicious commented:
Is it fine if I leave an approval just to unblock the CI? I feel like I don't have enough context on Discovery to give this a proper review.

kimlisa commented:
@ravicious @ryanclark i think it's fine to just leave a bot approval since we have two approvals for this PR. this is what i often do on teleport side after I get two approvals of my selection (i try to manually pick people with the most context and they sometimes don't satisfy the reviewers requirement)

kimlisa committed:
Merge branch 'master' into ryan/discover-remove-timers
